orderer/common/channelparticipation/restapi.go

```diff
@@ -50,23 +78,122 @@ func (h *HTTPHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	h.router.ServeHTTP(resp, req)
+}
+
+// List all channels
+func (h *HTTPHandler) serveListAll(resp http.ResponseWriter, req *http.Request) {
+	_, err := negotiateContentType(req) // Only application/json for now
+	if err != nil {
+		h.sendResponseJsonError(resp, http.StatusNotAcceptable, err)
+		return
+	}
+	channelList := h.registrar.ListAllChannels()
+	for i, info := range channelList.Channels {
+		channelList.Channels[i].URL = path.Join(URLBaseV1Channels, info.Name)
+	}
+	h.sendResponseOK(resp, channelList)
+}
+
+// List a single channel
+func (h *HTTPHandler) serveListOne(resp http.ResponseWriter, req *http.Request) {
+	_, err := negotiateContentType(req) // Only application/json for now
+	if err != nil {
+		h.sendResponseJsonError(resp, http.StatusNotAcceptable, err)
+		return
+	}
 
-	default:
-		err := errors.Errorf("invalid request method: %s", req.Method)
-		h.sendResponseJsonError(resp, http.StatusBadRequest, err)
+	channelID, ok := mux.Vars(req)[channelIDKey]
+	if !ok {
+		h.sendResponseJsonError(resp, http.StatusInternalServerError, errors.Wrap(err, "missing channel ID"))
+		return
 	}
+
+	if err = configtx.ValidateChannelID(channelID); err != nil {
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "invalid channel ID"))
+		return
+	}
+	infoFull, err := h.registrar.ListChannel(channelID)
+	if err != nil {
+		h.sendResponseJsonError(resp, http.StatusNotFound, err)
+		return
+	}
+	h.sendResponseOK(resp, infoFull)
+}
+
+// Join a channel
+func (h *HTTPHandler) serveJoin(resp http.ResponseWriter, req *http.Request) {
+	_, err := negotiateContentType(req) // Only application/json for now
+	if err != nil {
+		h.sendResponseJsonError(resp, http.StatusNotAcceptable, err)
+		return
+	}
+
+	//TODO
+	err = errors.Errorf("not implemented yet: %s %s", req.Method, req.URL.String())
+	h.sendResponseJsonError(resp, http.StatusNotImplemented, err)
+}
+
+// Remove a channel
+func (h *HTTPHandler) serveRemove(resp http.ResponseWriter, req *http.Request) {
+	_, err := negotiateContentType(req) // Only application/json for now
+	if err != nil {
+		h.sendResponseJsonError(resp, http.StatusNotAcceptable, err)
+		return
+	}
+
+	//TODO
+	err = errors.Errorf("not implemented yet: %s %s", req.Method, req.URL.String())
+	h.sendResponseJsonError(resp, http.StatusNotImplemented, err)
+}
+
+func (h *HTTPHandler) serveNotAllowed(resp http.ResponseWriter, req *http.Request) {
+	err := errors.Errorf("invalid request method: %s", req.Method)
+	encoder := json.NewEncoder(resp)
+	resp.WriteHeader(http.StatusMethodNotAllowed)
+	if _, ok := mux.Vars(req)[channelIDKey]; ok {
+		resp.Header().Set("Allow", "GET, POST, DELETE")
+	} else {
+		resp.Header().Set("Allow", "GET")
+	}
+	resp.Header().Set("Content-Type", "application/json")
+	if err := encoder.Encode(&types.ErrorResponse{Error: err.Error()}); err != nil {
+		h.logger.Errorf("failed to encode error, err: %s", err)
+	}
+}
+
+func negotiateContentType(req *http.Request) (string, error) {
```

Comment:
instead what you can do is to define headers you would accept while initializing gorilla/mux router, e.g.

```
    router.Headers("Content-Type", "application/json")
```

Reply:
Or there is an option to handle regexps as well


```
    router.HeadersRegexp("Content-Type", "application/(text|json)")
```

Reply:
Good observation. I will use it to ensure that the POST arrives with Content-type: application/json. (GET and DELETE have empty bodies, so there is no need to check.)

The Headers() matcher is good for filtering out requests that come in with bad Content-Type, for example, but here I am dealing with the Accept header. It needs special treatment that is not easy with regexp, and in particular it requires a response with http.StatusNotAcceptable and an error message in the body that describes what the server can deliver. In other cases it requires selecting the Mime type according to the q value of preferred types and so on. 

See also: https://github.com/gorilla/mux/issues/18